### DISCUSSION TITLE
how to process tsconfig

### DISCUSSION THREAD

**Discussion by robbiemu** [TypeStrong/ts-node#1260] [Q&A] [answered]
I am using ts-worker to run a node-worker .. it is a very lightweight wrapper that uses ts-node to compile before creating the worker.

In using this, I am getting the error:
```none
Error [ERR_UNHANDLED_ERROR]: Unhandled error. ({
  diagnosticText: "src/blockchain/utilities.ts(117,95): error TS2304: Cannot find name 'UUID'.\n",
  diagnosticCodes: [ 2304 ]
})
```
_see: https://github.com/viniciusestevam/ts-worker/issues/2_

`UUID` is defined in globals.d.ts: `type UUID = number | string`
This makes me think that ts-node is not respecting my tsconfig.json file. I see that one can call it like this:
```javascript
require('ts-node').register({ /* options */ })
```
and in the ts-worker library it is currently used like this:
```javascript
require('ts-node').register();
require(path.resolve(workerData.workerFilePath));
```
I would make a PR to handle this case, but I dont know how to get the tsconfig in there, or even if that is the right solution to my problem. 

edit: github version of this same issue, reproduced without relying on ts-worker: https://stackoverflow.com/questions/66430914/unable-to-access-path-shortcut-imports-in-worker

**Comment by cspotcode** [comment, marked as answer]
Try with `--files`.  ts-node has an optimization to speed up typechecking where it only checks files referenced directly or indirectly by your entrypoint script.  For example, if you run `ts-node ./index.ts` and `index.ts` imports `foo.ts` then it'll typecheck those 2 files.  But if you have types from another file, `bar.ts`, then those won't be pulled in.

`--files` disables this optimization and makes `ts-node` match your tsconfig's files array.

Where if UUID declared?  Is it a global type?  Or do you need to `import` it?

Also, you should be able to use `ts-node` in worker threads like this:

```
NODE_OPTIONS='-r ts-node/register' TS_NODE_PROJECT=path/to/tsconfig.json node ./your-script.ts
```

**Reply by cspotcode** [reply]
Also, I have implemented a `--showConfig` flag (#1243) which will be published in the next release of `ts-node`.  It mimics `tsc --showConfig` but also includes ts-node options.  I added it to make it easier for users to understand precisely what configuration `ts-node` is using and to diagnose configuration issues.  If you're feeling adventurous, you can try it out today:

```
npm install TypeStrong/ts-node#main
```

**Reply by robbiemu** [reply]
it is globally defined in globals.d.ts in the types/ directory.. my tsconfig has this section that would normally pull it in:
```json
  "include": [
    "src/**/*",
    "types/*.d.ts",
  ],
```
I think your comments speak exclusively to commandline usage. Otherwise, how do I pass `--files` in this use case? like this??
```javascript
require('ts-node').register({ '--files': true });
require(path.resolve(workerData.workerFilePath));

```

**Reply by cspotcode** [reply]
When passing it programmatically, you need to obey our TypeScript type declaration which will tell you the correct name for every option you can pass to `.register()`

Options can also be placed into your tsconfig.json.  Here is an example: https://typestrong.org/ts-node/docs/configuration
The options you can place into your tsconfig.json are governed by the JSON schema which is also linked on that page.

**Reply by cspotcode** [reply]
I recommend checking out the new `--showConfig` option because it will show you how using `--files` and not using `--files` affects the set of files pulled by the compiler.  With `--files`, we match the behavior of `tsc` and pull in everything matched by `"include"`.  Without `--files`, you'll see that we behave differently and only pull in your entry-point script, which is then responsible for referencing all other files directly or indirectly.  If nothing references your `globals.d.ts` via either an `import` or a `/// <reference>` then those types will be unavailable.

**Reply by robbiemu** [reply]
thank you

**Reply by robbiemu** [reply]
in fact this works for the globals, but the typescript paths are not working. I am getting:

```none
events.js:291
      throw er; // Unhandled 'error' event
      ^
Error: Cannot find module '@lib/assert'
```

If I preprocess the paths array before passing it on to the _register_ call, I get the same error. 

I also migrated from ts-worker, using a local worker.js and worker.ts file.

**Reply by cspotcode** [reply]
That sounds like a separate issue, and I created a dedicated answer for it: #1263
